### ISSUE TITLE
yubikey-agent blocks Yubikey Manager

### ISSUE THREAD

**Issue by joneskoo** [FiloSottile/yubikey-agent#4]
At least on macOS, if I have yubikey-agent running and I have authenticated with it (just starting yubikey-agent does not reproduce the problem), Yubikey Manager hangs. It appears to hang in a call to PCSC, and with some experiments the PIV application seems to be where it blocks - which makes sense obviously.

I'm creating this to make a reminder now to document this behavior in upcoming README (including how to stop the agent if Yubikey Manager is needed). Depending on how setup will work, maybe the Yubikey Manager is not needed by general audience but there needs to be a way to free the interface (easiest to stop the agent). Feel free to close/replace with some other TODO of course.

**Comment by FiloSottile** [comment]
Yup, same with gpg-agent, seems to be a PCSC limitation. We'll definitely document this. There's also go-piv/piv-go#47 upstream to look into ways to make this better.

The interface is released on SIGHUP, so you can use `killall -HUP yubikey-agent`.

**Comment by ryan-gerstenkorn-sp** [comment]
Just FYI, this workaround seems to fix a similar issue when using `ykman oath code --single ...` as well.

Just so it's searchable here's the error you get when trying this when yubikey-agent is running:
```
Error: Failed connecting to YubiKey 5 [FIDO+CCID]. Make sure the application have                     the required permissions.
```

**Comment by FiloSottile** [comment]
Interestingly, my YubiKey 5 seems to persist the PIN cache across sessions, and even yubikey-agent restarts (but not unplug-replug cycles, as expected). In this case it would be far more acceptable to just drop the session every time.

I bet using a different applet will still trash the PIN cache, but that's probably ok.

**Comment by FiloSottile** [comment]
Huh, it might be that PIN caching has always spanned sessions, but on older firmwares getting the serial number requires switching applets, so effectively we were dropping it at every session. Maybe there's an argument for not taking an exclusive lock, and just not reading the serial on older firmwares.

**Comment by WhyNotHugo** [comment]
Hi! I want to use `yubikey-agent` for a very few SSH keys. Mostly, the ones for sites that don't support `ed25519-sk` keys.

This issue is a a big pain, especially because I use the yubikey as a FIDO device a lot too as well as for GPG.

I wouldn't mind having to type the PIN each time I use yubikey-agent if that's the price for it to not lock my yubikey the rest of the time.

Is this approach somehow possible? I think it could be a useful solution for some until upstream unblocks this issue.

**Comment by smlx** [comment]
Wouldn't #44 essentially solve this? `launchd` also supports socket activation. Here's the `launchd` config and associated socket activation code I use in `piv-agent`. It's only lightly tested on macOS since I don't use that OS regularly, but it does work:
https://github.com/smlx/piv-agent/blob/main/deploy/launchd/com.github.smlx.piv-agent.plist
https://github.com/smlx/piv-agent/blob/main/internal/sockets/get_darwin.go

**Comment by ezekiel** [comment]
I may not fully understand the context here, but it might be more worthwhile overall to place some efforts on this PR on the upstream `go-piv` library: https://github.com/go-piv/piv-go/pull/100

This will allow yubikey-agent and piv-agent to `Open` _SHARED_ access to the key, and avoid blocking other applications which do the same ( like ssh-agent, p11-kit, etc ).

**Comment by cedws** [comment]
https://github.com/go-piv/piv-go/issues/47#issuecomment-1280781982

>Looks like the PR for piv-go is stuck.

>Would it make sense to build a broker that could be used by this project and others? Though it would be difficult to get projects to adopt it.

**Comment by jdeepwell** [comment]
Should commit 2e5376c solve this (at least with Yubikeys version >=5? I am running yubikey-agent 0.1.6 (via homebrew) on macOS 14.7.2 (23H311) and as soon as any application is opening an SSH connection and I enter the PIN, the Yubikey is unaccessible to e.g. smimesign – which makes it unusable, as I am opening the git/SSH connection to actually *do* a git commit and push, which I then can't.

**Comment by h4fnp** [comment]
I ran into this as well and wrote a small helper script to switch between OpenPGP (gpg-agent) and SSH (ssh-agent + yubikey-agent) modes on Linux with systemd user services. It’s just a personal utility; I make no copyright claims and anyone is free to use, redistribute and share it at their own risk: [https://gist.github.com/h4fnp/2eaf4d790dc52fe2d9fc21c1f9ea91b7](https://gist.github.com/h4fnp/2eaf4d790dc52fe2d9fc21c1f9ea91b7)

It might help others who want a simple toggle/status workflow for YubiKey modes.

**Comment by h4fnp** [comment]
Follow-up to my earlier comment: I've since replaced the mode-toggle script with a setup where both agents stay running and the card is handed over **automatically**, in both directions. No toggling, and the PIN stays cached on each side. Verified on Debian (GnuPG 2.4.9,
  yubikey-agent 0.1.6, key in PIV slot 9a + OpenPGP applet for pass).
  
  The trick is two release hooks that fire exactly when each side is about to need the card:
  
  **GPG → SSH**: in `~/.ssh/config`, before the first `Host` block:

  ``` 
  Match exec "gpgconf --kill scdaemon"
  ```
  
  ssh evaluates this while parsing the config, i.e. before authentication — and everything that speaks SSH (scp, sftp, git, rsync, ansible, …) parses this file. Killing scdaemon releases the card; gpg-agent respawns it on demand.
  
  **SSH → GPG**: start gpg-agent with `--scdaemon-program` pointing at a tiny wrapper:

  ```bash
  #!/usr/bin/env bash
  killall -q -HUP yubikey-agent
  exec /usr/lib/gnupg/scdaemon "$@"
  ```

  Because the SSH side always kills scdaemon, any GPG use after SSH activity *must* respawn scdaemon — which routes through the wrapper and frees the card at exactly the right moment. Consecutive GPG operations reuse the running scdaemon, so the card session (and PIN cache)  survives. With systemd user units, a drop-in keeps it portable across machines (`%h` = home):

  ```ini
  # ~/.config/systemd/user/gpg-agent.service.d/override.conf
  [Service]
  ExecStart=
  ExecStart=/usr/bin/gpg-agent --supervised --scdaemon-program %h/.local/bin/scdaemon
  ```
 
  Plus `disable-ccid` in `~/.gnupg/scdaemon.conf` so scdaemon goes through pcscd. (I initially also had `pcsc-shared`, but with this handover it's unnecessary — and dropping it restores GPG PIN caching across operations.)
 
  Result: consecutive `ssh` = one PIN, consecutive `gpg`/`pass` = one PIN; only alternating between the two re-prompts, which is inherent (each handover drops the other side's cache). Even an ansible playbook that authenticates over SSH and then decrypts pass secrets mid-run
  works, since ansible's multiplexed connections survive the HUP.
